subsys/bluetooth/mesh/gen_lvl_srv.c

```diff
@@ -0,0 +1,208 @@
+{
+	struct gen_lvl_srv *srv = model->user_data;
+	struct gen_lvl_status status = { 0 };
+	struct model_transition transition;
+	struct gen_lvl_set set;
+
+	set.level = net_buf_simple_pull_le16(buf);
+	set.new_transaction = tid_check_and_update(
+		&srv->tid, net_buf_simple_pull_u8(buf), ctx);
+	set.transition = transition_get(srv, &transition, buf);
+
+	srv->handlers->set(srv, ctx, &set, &status);
+
+	if (ack) {
+		rsp_status(model, ctx, &status);
+	}
+
+	(void)gen_lvl_srv_pub(srv, NULL, &status);
+}
+
+static void delta_set(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+		      struct net_buf_simple *buf, bool ack)
+{
+	struct gen_lvl_srv *srv = model->user_data;
+	struct gen_lvl_status status = { 0 };
+	struct gen_lvl_delta_set delta_set;
+	struct model_transition transition;
+
+	delta_set.delta = net_buf_simple_pull_le32(buf);
+	delta_set.new_transaction = tid_check_and_update(
+		&srv->tid, net_buf_simple_pull_u8(buf), ctx);
+	delta_set.transition = transition_get(srv, &transition, buf);
+
+	srv->handlers->delta_set(srv, ctx, &delta_set, &status);
+
+	if (ack) {
+		rsp_status(model, ctx, &status);
+	}
+
+	(void)gen_lvl_srv_pub(srv, NULL, &status);
+}
+
+static void move_set(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+		     struct net_buf_simple *buf, bool ack)
+{
+	struct gen_lvl_srv *srv = model->user_data;
+	struct gen_lvl_status status = { 0 };
+	struct model_transition transition;
+	struct gen_lvl_move_set move_set;
+
+	move_set.delta = net_buf_simple_pull_le16(buf);
+	move_set.new_transaction = tid_check_and_update(
+		&srv->tid, net_buf_simple_pull_u8(buf), ctx);
+	move_set.transition = transition_get(srv, &transition, buf);
+
+	srv->handlers->move_set(srv, ctx, &move_set, &status);
+
+	if (ack) {
+		rsp_status(model, ctx, &status);
+	}
+
+	(void)gen_lvl_srv_pub(srv, ctx, &status);
+}
+
+/* Message handlers */
+
+static void handle_set(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+		       struct net_buf_simple *buf)
+{
+	set(model, ctx, buf, true);
+}
+
+static void handle_set_unack(struct bt_mesh_model *model,
+			     struct bt_mesh_msg_ctx *ctx,
+			     struct net_buf_simple *buf)
+{
+	set(model, ctx, buf, false);
+}
+
+static void handle_delta_set(struct bt_mesh_model *model,
+			     struct bt_mesh_msg_ctx *ctx,
+			     struct net_buf_simple *buf)
+{
+	delta_set(model, ctx, buf, true);
+}
+
+static void handle_delta_set_unack(struct bt_mesh_model *model,
+				   struct bt_mesh_msg_ctx *ctx,
+				   struct net_buf_simple *buf)
+{
+	delta_set(model, ctx, buf, false);
+}
+
+static void handle_move_set(struct bt_mesh_model *model,
+			    struct bt_mesh_msg_ctx *ctx,
+			    struct net_buf_simple *buf)
+{
+	move_set(model, ctx, buf, true);
+}
+
+static void handle_move_set_unack(struct bt_mesh_model *model,
+				  struct bt_mesh_msg_ctx *ctx,
+				  struct net_buf_simple *buf)
+{
+	move_set(model, ctx, buf, false);
+}
+
+const struct bt_mesh_model_op _gen_lvl_srv_op[] = {
```

Comment:
`min_len` includes the opcode size as well.

Reply:
It shouldn't. The opcode is pulled out of the buffer before the min_len check in access.c, and is therefore not part of the calculation.